jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java

```diff
@@ -0,0 +1,445 @@
+                        // extend the path if the extension is still negative and correctly labeled
+                        double edgeWeight = graph.getEdgeWeight(e);
+                        int currentLabel = labelMap.get(currentVertex);
+                        if (!path.getLabels().contains(currentLabel) && path.getCost() + edgeWeight < 0) {
+                            LabeledPath<V> newPath = path.clone();
+                            newPath.addVertex(currentVertex, edgeWeight, currentLabel);
+
+                            /*
+                             * check if paths are dominated, i.e. if the path is definitely worse than other paths
+                             * and does not have to be considered in the future
+                             */
+                            if (!removeDominatedPathsOfLengthKplus1(newPath, pathsLengthKplus1)) {
+                                if(!removeDominatedPathsOfLengthK(newPath, pathsLengthK)) {
+                                    addPathToPathStructure(pathsLengthKplus1, newPath);
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            // update k and the corresponding sets
+            k += 1;
+            pathsLengthK = pathsLengthKplus1;
+            pathsLengthKplus1 = new LinkedHashMap<>();
+        }
+
+        return new GraphWalk<>(graph, new ArrayList<>(lengthBound), Double.MAX_VALUE);
+    }
+
+    /**
+     * Removes all paths that are dominated from all calculated paths of length k + 1.
+     * This is important out of efficiency reasons, otherwise many unnecessary paths may
+     * be considered in further calculations.
+     *
+     * @param path the currently calculated path
+     * @param pathsLengthKplus1 all before calculated paths of length k + 1
+     *
+     * @return whether <code>path</code> was removed in the procedure
+     */
+    private boolean removeDominatedPathsOfLengthKplus1(LabeledPath<V> path, Map<PathSetKey<V>, TreeSet<LabeledPath<V>>> pathsLengthKplus1) {
+        boolean removePath = false;
+
+        TreeSet<LabeledPath<V>> pathsToCheck = pathsLengthKplus1.get(PathSetKey.of(path.getHead(), path.getTail(), path.getLabels()));
+        if(pathsToCheck != null) {
+            // simulates domination test by using the index structure
+            for (Iterator<LabeledPath<V>> it = pathsToCheck.descendingIterator(); it.hasNext(); ) {
+                LabeledPath<V> pathInSetKplus1 = it.next();
+
+                if (pathInSetKplus1.getCost() < path.getCost()) {
+                    removePath = true;
+                    /*
+                     * we can break because domination of paths is transitive, i.e. pathInSetKplus1 already removed the dominated paths
+                     */
+                    break;
+                }
+                if (path.getCost() < pathInSetKplus1.getCost()) {
```

Comment:
I concur that calling clear() on a submap view is probably no better than iteration (I think under the covers that is how it ends up implemented).

There is one other way to do it though, which is to replace the original TreeSet with the subset to retain.  The implementation for this is efficient because it just keeps begin/end pointers into the original set (and doing this more than once just moves the pointers).

But stepping back for a moment, is there even a need to keep a TreeSet at all?  In the end, don't we only keep the path of lowest cost?